### PR TITLE
Problem: we don't have any tests for malicious strings

### PR DESCRIPTION
**Pull request by z-bowen**
## Solution

Use a parameterized test that fuzzes over a library of potentially hazardous strings. I'm using the 'big_list_of_naughty_strings', which includes some nice sql-injections, some crazy unicode sequences, and some invocations to elder gods... This up shot is, the work of maintaining the 'naughty strings' is off-loaded, and the test file generates individual tests for each string, so we should stay compliant if any new attack vectors open up.

## Issues Resolved

Contributing as part of the stability push for release of beta

## BEPs Implemented

NA

### CODE REVIEW

File: Dockerfile-dev [hunk 1]
```diff
@@ -33,5 +33,6 @@ RUN mkdir -p /usr/src/app
 COPY . /usr/src/app/
 WORKDIR /usr/src/app
 RUN pip install --no-cache-dir --process-dependency-links -e .[dev]
-RUN bigchaindb -y configure "$backend"
+RUN pip install blns
```
[the remark this answers is not in the document]
Comment (z-bowen) [reply Dockerfile-dev]: I tested that, adding `blns` to both `dev_require` and `tests_require`, but in both cases, without the changes to `acceptance/python/Dockerfile`, I can't get the import to work in a new Docker image.

File: Dockerfile-dev [hunk 2]
```diff
@@ -33,5 +33,5 @@ RUN mkdir -p /usr/src/app
 COPY . /usr/src/app/
 WORKDIR /usr/src/app
 RUN pip install --no-cache-dir --process-dependency-links -e .[dev]
-RUN bigchaindb -y configure "$backend"
+RUN bigchaindb -y configure
```
Comment (ttmc) [inline Dockerfile-dev]: I generally follow the advice, "If it ain't broke, don't fix it." so I wouldn't have made this change or the corresponding one in `docker-compose.yml`. We might add support for other backends in the future.

File: acceptance/python/src/test_naughty_strings.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        assert status_code == 400, status_code
+        assert re.fullmatch(regex, error), error
+    else:
+        raise TypeError(sent_transaction)
+
+
+@pytest.mark.parametrize("naughty_string", naughty_strings, ids=naughty_strings)
+def test_naughty_keys(naughty_string):
+
+    asset = {'data': {naughty_string: 'naughty_value'}}
```
Comment (ttmc) [inline acceptance/python/src/test_naughty_strings.py]: This is a minor quibble about semantics...

Here, the value is `'naughty_value'`, which is actually just a normal, non-strange string, a non-naughty value. In other words, `'naughty_value'` is not a naughty value, and this melts my brain a little bit.

treachery

Maybe replace `'naughty_value'` with `'normal_value'`

Similar comments can be made for the following line, and for the `'naughty_key'` in the `test_naughty_values()` function.

File: acceptance/python/src/test_naughty_strings.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    # Then she checks that either her transaction was written to the block chain,
+    # or she got back a nice error message
+    if type(sent_transaction) == dict:
+        if 'id' in sent_transaction.keys():
+            tx_id = sent_transaction['id']
+            assert bdb.transactions.retrieve(tx_id), sent_transaction
+    elif type(sent_transaction) == BadRequest:
+        status_code = sent_transaction.status_code
+        error = sent_transaction.error
+        regex = '\{"message":"Invalid transaction \\(ValidationError\\): Invalid key name .* in asset object. The key name cannot contain characters .* or null characters","status":400\}\n'
```
Comment (ttmc) [inline acceptance/python/src/test_naughty_strings.py]: Also, a key can't start with `$`. I wonder why that case is not mentioned.

File: Dockerfile-dev (hunk 2 above)
Comment (z-bowen) [reply Dockerfile-dev]: This does, indeed, break the test. Removing this one change causes the execution to fail while building the Docker images. I'm not totally clear on why this happens, but unless removing the `"$backend"` breaks something else, I think we should pull it.

File: acceptance/python/src/test_naughty_strings.py (hunk 3 above)
Comment (z-bowen) [reply acceptance/python/src/test_naughty_strings.py]: Fixed

File: Dockerfile-dev (hunk 2 above)
Comment (ttmc) [reply Dockerfile-dev]: Oh, interesting! I guess you should keep this change then.

**Comment by kansi** [comment]
@z-bowen Why was this PR closed?

**Pull request closed without merging**